daemon/libnetwork/network.go

```diff
@@ -1521,123 +1526,129 @@ func (n *Network) ipamAllocate() (retErr error) {
 				}
 			}
 		}()
 
+		d := &IpamInfo{
+			PoolID: alloc.PoolID,
+			Meta:   alloc.Meta,
+			IPAMData: driverapi.IPAMData{
+				Pool:         netiputil.ToIPNet(alloc.Pool),
+				AddressSpace: n.addrSpace,
+			},
+		}
+		newInfo = append(newInfo, d)
+
+		// During network restore, if the original network-create request did not specify a
+		// gateway, use the previously allocated gateway.
+		prefGateway := cfg.Gateway
+		if prefGateway == "" && len(ipamInfo) > i && ipamInfo[i].Gateway != nil {
+			prefGateway = ipamInfo[i].Gateway.IP.String()
+		}
+
 		// If there's no user-configured gateway address but the IPAM driver returned a gw when it
 		// set up the pool, use it. (It doesn't need to be requested/reserved in IPAM.)
-		if cfg.Gateway == "" {
+		if prefGateway == "" {
 			if gws, ok := d.Meta[netlabel.Gateway]; ok {
 				if d.Gateway, err = types.ParseCIDR(gws); err != nil {
-					return types.InvalidParameterErrorf("failed to parse gateway address (%v) returned by ipam driver: %v", gws, err)
+					return nil, types.InvalidParameterErrorf("failed to parse gateway address (%v) returned by ipam driver: %v", gws, err)
 				}
 			}
 		}
 
 		// If there's still no gateway, reserve cfg.Gateway if the user specified it. Else,
 		// if the driver wants a gateway, let the IPAM driver select an address.
-		if d.Gateway == nil && (cfg.Gateway != "" || !skipGwAlloc) {
+		if d.Gateway == nil && (prefGateway != "" || !skipGwAlloc) {
 			gatewayOpts := map[string]string{
 				ipamapi.RequestAddressType: netlabel.Gateway,
 			}
-			if d.Gateway, _, err = ipam.RequestAddress(d.PoolID, net.ParseIP(cfg.Gateway), gatewayOpts); err != nil {
-				return types.InternalErrorf("failed to allocate gateway (%v): %v", cfg.Gateway, err)
+			if d.Gateway, _, err = ipam.RequestAddress(d.PoolID, net.ParseIP(prefGateway), gatewayOpts); err != nil {
+				return nil, types.InternalErrorf("failed to allocate gateway (%v): %v", prefGateway, err)
 			}
 		}
 
-		// Auxiliary addresses must be part of the master address pool
-		// If they fall into the container addressable pool, libnetwork will reserve them
+		// Auxiliary addresses must be part of the master address pool.
+		// If they fall into the container addressable pool, reserve them.
 		if cfg.AuxAddresses != nil {
 			var ip net.IP
 			d.IPAMData.AuxAddresses = make(map[string]*net.IPNet, len(cfg.AuxAddresses))
 			for k, v := range cfg.AuxAddresses {
 				if ip = net.ParseIP(v); ip == nil {
-					return types.InvalidParameterErrorf("non parsable secondary ip address (%s:%s) passed for network %s", k, v, n.Name())
+					return nil, types.InvalidParameterErrorf("non parsable secondary ip address (%s:%s) passed for network %s", k, v, n.Name())
 				}
 				if !d.Pool.Contains(ip) {
-					return types.ForbiddenErrorf("auxiliary address: (%s:%s) must belong to the master pool: %s", k, v, d.Pool)
+					return nil, types.ForbiddenErrorf("auxiliary address: (%s:%s) must belong to the master pool: %s", k, v, d.Pool)
 				}
```

Comment:
Not necessarily for this PR, but I think we should move this to `cfg.Validate()` and disallow `AuxAddresses` if no specific `Subnet` is provided.